crates/agent/src/lib.rs

```diff
@@ -106,3 +203,175 @@ async fn upsert_draft_specs(
+      ),
+      p4 as (
+        delete from draft_specs where id = '1111000000000000'
+      ),
+      p3 as (
+          delete from live_specs where id = '1000000000000000'
+      ),
+      p2 as (
+        delete from drafts where id = '1110000000000000'
+      ),
+      p1 as (
+        delete from auth.users where id = '43a18a3e-5a59-11ed-9b6a-0242ac120002'
+      )
+      select 1;
+    "#;
+
+    #[derive(Debug)]
+    struct MockHandler {
+        notifier: tokio::sync::mpsc::UnboundedSender<()>,
+        table_name: &'static str,
+    }
+
+    impl MockHandler {
+        fn new(table_name: &'static str, notifier: tokio::sync::mpsc::UnboundedSender<()>) -> Self {
+            MockHandler {
+                notifier,
+                table_name,
+            }
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl Handler for MockHandler {
+        async fn handle(&mut self, _: &sqlx::PgPool) -> anyhow::Result<HandlerStatus> {
+            self.notifier.send(()).unwrap();
+            Ok(HandlerStatus::Idle)
+        }
+
+        fn table_name(&self) -> &'static str {
+            &self.table_name
+        }
+    }
+
+    #[tokio::test]
+    async fn test_handlers_react_quickly() {
+        let pg_pool = PgPool::connect(&FIXED_DATABASE_URL).await.unwrap();
+
+        let (handler_notify_tx, mut handler_notify_rx) =
+            tokio::sync::mpsc::unbounded_channel::<()>();
+        // Must allow `exit_tx` to exist here,
+        // otherwise it'll get instantly dropped and kill the server prematurely
+        #[allow(unused_variables)]
+        let (exit_tx, exit_rx) = tokio::sync::oneshot::channel::<()>();
+
+        let server = serve(
+            vec![Box::new(MockHandler::new(
+                "publications",
+                handler_notify_tx,
+            ))],
+            pg_pool.clone(),
+            exit_rx.map(|_| ()),
+        );
+
+        tokio::pin!(server);
+
+        tokio::select! {
+            res = &mut server => {
+                Err(anyhow::anyhow!("Handler unexpectedly exited: {:#?}", res))
+            }
+            res = async move {
+                // Do this 10 times in a row to make sure that our handler gets called consistently quickly
+                for _ in 0..10{
+                    let mut txn = pg_pool.begin().await.unwrap();
+                    // Sets up the database to have a valid publication task and associated draft/specs
+                    sqlx::query(include_str!("publications/test_resources/happy_path.sql"))
+                        .execute(&mut txn)
+                        .await
+                        .unwrap();
+
+                    // We have to commit the transaction for the NOTIFY to get sent
+                    txn.commit().await.unwrap();
+
+                    // Make sure that our mock publication handler was called within 25ms
+                    let res = tokio::time::timeout(Duration::from_millis(25), handler_notify_rx.recv().map(|_|()))
```

Comment:
this is likely to flake on CI. Do we _need_ to bound how long it will take? It should be enough to just assert that it's delivered as expected.

Reply:
The intent was to prove that it wasn't polling, and instead would react "immediately"... but I do see your point about flaking in CI. I'll remove the timeout, as it's pretty clear from reading the code that it's not polling any more :P